crates/core/src/transcript/proof.rs

```diff
@@ -208,48 +208,79 @@ impl<'a> TranscriptProofBuilder<'a> {
 
                 let dir_idx = (direction, idx);
 
-                if !encoding_tree.contains(&dir_idx) {
-                    return Err(TranscriptProofBuilderError::new(
-                        BuilderErrorKind::MissingCommitment,
-                        format!(
-                            "encoding commitment is missing for ranges in {} transcript",
-                            direction
-                        ),
-                    ));
+                // Insert the rangeset if it's in the encoding tree (which means it's a
+                // committed rangeset).
+                if encoding_tree.contains(&dir_idx) {
+                    self.encoding_proof_idxs.insert(dir_idx);
+                } else {
+                    let mut missing_commitment = true;
+                    // Check if there is any committed rangeset in the encoding tree that is a
+                    // subset of the rangeset — if yes, insert them.
+                    for committed_dir_idx in encoding_tree
+                        .transcript_indices()
+                        .into_iter()
+                        .filter(|(dir, _)| *dir == dir_idx.0)
+                    {
+                        if committed_dir_idx.1.is_subset(&dir_idx.1) {
+                            self.encoding_proof_idxs.insert(committed_dir_idx.clone());
```

Comment:
I didnt review yet, overall the approach is correct, imo, but
we need to make sure that what we are inserting here is not a subset of what is already present in `encoding_proof_idxs`.
Suppose we already inserted 0..100 into `encoding_proof_idxs`. We don't want to insert anything that is a subset of 0..100 since that would unnecessarily increase the proof size and verification time. 

@yuroitaki

Reply:
That's a good point — though I believe that problem already exists before this PR? e.g. where user first passes in a committed header range (incl. key and value), and then only the header key range; should this be taken care in a different PR?

Reply:
I would do it in this PR.
Are you saying that there was code in the examples where the user first calls reveal on key+value and then calls reveal on the key? That seems redundant.

Reply:
> I would do it in this PR. Are you saying that there was code in the examples where the user first calls `reveal` on `key+value` and then calls `reveal` on the key? That seems redundant.

Ah no code in examples — I was just giving a scenario how this issue theoretically already exists currently (before this PR); because to pass in a subset of what's present in `encoding_proof idxs`, it will mean

1. User first pass in a rangeset, then a subset rangeset (both are/contain committed range(set)) [already possible before this PR]
2. User first pass in a rangeset, then another rangeset that intersects with the former, where the intersection contains committed range(set) [possible after this PR]

I only thought of (1) in my previous comment, and as you said they seemed to be redundant (and unlikely to happen) because it means e.g. user first calls reveal on `key+value` and then calls reveal on the `key`, or user first calls reveal on `header.start..body.end` and then `header.start..header.end`; 

Though now I think about it, (1) might happen more likely after this PR, e.g. first reveal `request.without_data().start..request.without_data().end`, then reveal `request.request.target` — makes sense to look into the check you suggested

Reply:
Thanks for explaining. I see, in the scenario 1) before this PR there would be a secong encoding proof added which is a redundancy that we overlooked.

1) should be permissible behavior. If the user tries to reveal a subset of what already was revealed, it should be a no-op.

Reply:
> > I would do it in this PR. Are you saying that there was code in the examples where the user first calls `reveal` on `key+value` and then calls `reveal` on the key? That seems redundant.
> 
> Ah no code in examples — I was just giving a scenario how this issue theoretically already exists currently (before this PR); because to pass in a subset of what's present in `encoding_proof idxs`, it will mean
> 
> 1. User first pass in a rangeset, then a subset rangeset (both are/contain committed range(set)) [already possible before this PR]
> 2. User first pass in a rangeset, then another rangeset that intersects with the former, where the intersection contains committed range(set) [possible after this PR]
> 
> I only thought of (1) in my previous comment, and they seemed to be redundant (and unlikely to happen) because it means e.g. user first calls reveal on `key+value` and then calls reveal on the `key`, or user first calls reveal on `header.start..body.end` and then `header.start..header.end`;
> 
> Though now I think about it, (1) might happen more likely after this PR, e.g. first reveal `request.without_data().start..request.without_data().end`, then reveal `request.request.target` — makes sense to look into the check you suggested

@themighty1 I realised what I said here is mostly wrong and only partially right — when user reveals superset rangesets (union of multiple committed rangesets), (1) and (2) will not insert a subset of what's present in `encoding_proof_idxs`. Instead, the following conditions will:

(1) Among the committed rangesets, there are rangesets that are subset of another; this is common, e.g. `header key+value` and `header key`.
(2) User reveals a superset rangeset (call it `S`), which covers both `header key+value` and `header key`, causing both to be inserted into `encoding_proof_idxs`.

Subsequent revealing of another superset rangeset that is (1) subset of `S` or  (2) intersects `S`, will not insert `header key+value` and `header key` into `encoding_proof_idxs` because both of them are already present in `encoding_proof_idxs`, which is a `HashMap`.

Reply:
For that, I have added a subset check now — https://github.com/tlsnotary/tlsn/pull/664/commits/85fb7c67c8acb3d7bb302318f28ac9c3fa562108; though I will need to do the same for hash-based commitments.